Assets/Tests/Runtime/ConnectionManagementTests.cs

```diff
@@ -0,0 +1,514 @@
+                var clientId = i;
+                yield return new WaitWhile(() => m_ClientNetworkManagers[clientId].ShutdownInProgress);
+                Assert.IsFalse(m_ClientNetworkManagers[clientId].IsConnectedClient, $"Client{clientId} has not shut down properly after losing connection.");
+            }
+
+            // Waiting for clients to fail to automatically reconnect
+            for (var i = 0; i < NumberOfClients; i++)
+            {
+                // expecting that error twice per client since there are two reconnect attempts
+                LogAssert.Expect(LogType.Error, k_FailedToConnectToServerErrorMessage);
+                LogAssert.Expect(LogType.Error, k_FailedToConnectToServerErrorMessage);
+            }
+            yield return WaitForClientsConnectedOrTimeOut();
+            for (var i = 0; i < NumberOfClients; i++)
+            {
+                Assert.IsFalse(m_ClientNetworkManagers[i].IsConnectedClient, $"Client{i} is connected while no server is running.");
+            }
+
+            Assert.AreEqual(NumberOfClients, nbReconnectingMsgsReceived, "Not all clients received a Reconnecting message.");
+            subscriptions.Dispose();
+        }
+
+        [UnityTest]
+        public IEnumerator ClientAndHostChangingRolesBetweenSessions_Success()
+        {
+            SetUniqueProfilesForEachClient();
+
+            StartHost();
+            AssertHostIsListening();
+
+            yield return ConnectClients();
+            AssertAllClientsAreConnected();
+
+            // Requesting shutdown on the server
+            m_ServerConnectionManager.RequestShutdown();
+            yield return new WaitWhile(() => m_ServerNetworkManager.IsListening);
+            Assert.IsFalse(m_ServerNetworkManager.IsHost);
+
+            // Waiting until shutdown is complete for the clients
+            for (var i = 0; i < NumberOfClients; i++)
+            {
+                var clientId = i;
+                yield return new WaitWhile(() => m_ClientNetworkManagers[clientId].ShutdownInProgress);
+                Assert.IsFalse(m_ClientNetworkManagers[clientId].IsConnectedClient, $"Client{clientId} has not shut down properly after losing connection.");
+            }
+
+            // Switching references for Client0 and Server temporarily
+            (m_ServerNetworkManager, m_ClientNetworkManagers[0]) = (m_ClientNetworkManagers[0], m_ServerNetworkManager);
+            (m_ServerConnectionManager, m_ClientConnectionManagers[0]) = (m_ClientConnectionManagers[0], m_ServerConnectionManager);
+
+            // recreate player prefab here
+            CreatePlayerPrefab();
+
+            StartHost();
+            AssertHostIsListening();
+
+            yield return ConnectClients();
+            AssertAllClientsAreConnected();
+
+            // Switching back references for Client0 and Server
+            (m_ServerNetworkManager, m_ClientNetworkManagers[0]) = (m_ClientNetworkManagers[0], m_ServerNetworkManager);
```

Comment:
See if this is needed to revert back roles since you also shutdown all clients as well as host.

Reply:
Not needed anymore, will remove